src/zookeeper.ts

```diff
@@ -1,316 +1,432 @@
-    });
+        if (this.locationPool) {
+            this.resolveLocationRequest(locator);
+        } else {
+            this.locationRequestQueue.push(locator);
+        }
+
+        if (this.locatorTimeout) {
+            setTimeout((() => {
+                if (locator.promise.isPending()) {
+                    locator.reject(new Error("ZOOKEEPER_TIMEOUT"));
+                }
+            }), this.locatorTimeout);
+        }
 
-    Promise.all(promises)
-        .then((newPool) => {
-          setPool(newPool.filter(Boolean));
-          emitter.emit(LocatorException.CODE["NEW_POOL"], path, pool);
-          processQueue();
-        })
-        .done();
-  }
-
-  function onChildrenChange(event : Event) {
-    emitter.emit(LocatorException.CODE["CHILDREN_CHANGED"], path, event);
-    clientWrapper.client.getChildren(path, onChildrenChange, onGetChildren);
-  }
-
-  function onExists(error : Error, stat : zookeeper.Stat) {
-    if (stat) {
-      stale = false;
-      emitter.emit(LocatorException.CODE["PATH_FOUND"], path);
-      clientWrapper.client.getChildren(path, onChildrenChange, onGetChildren);
-    } else {
-      stale = true;
-      emitter.emit(LocatorException.CODE["PATH_NOT_FOUND"], path);
-      setPool([]);
-      processQueue();
+        return locator.promise;
+    };
+
+    constructor (clientWrapper : ZookeeperClientWrapper,
+                 path : string,
+                 emitter : EventEmitter,
+                 dataExtractor : DataExtractor,
+                 locatorTimeout : number,
+                 strict : boolean) {
+        this.clientWrapper = clientWrapper;
+        this.path = path;
+        this.emitter = emitter;
+        this.dataExtractor = dataExtractor;
+        this.locatorTimeout = locatorTimeout;
+        this.strict = strict;
+        this.next = -1;
+        this.locationPool = null;
+        this.cachedLocationPool = null;
+        this.locationRequestQueue = [];
+        this.stale = false;
+
+        clientWrapper.on('NEW_CLIENT', () => {
+            this.setLocationPool(null);
+            clientWrapper.client.exists(path, this.onLocationPathExists);
+        });
+
+        if (clientWrapper.client) {
+            clientWrapper.client.exists(path, this.onLocationPathExists);
+        }
     }
-  }
-
-  clientWrapper.emitter.on('NEW_CLIENT', () => {
-    setPool(null);
-    clientWrapper.client.exists(path, onExists);
-  });
-
-  if (clientWrapper.client) {
-    clientWrapper.client.exists(path, onExists);
-  }
-
-  return function () {
-    var deferred = Promise.defer<Location>();
-    if (stale) {
-      setPool(null);
-      clientWrapper.client.exists(path, onExists);
+
+    /**
+     * update the list of locations available at this path
+     * @param newPool
+     */
+    private setLocationPool(newPool : Location[]) {
+        this.locationPool = newPool;
+        if (!this.cachedLocationPool || (newPool && newPool.length)) {
+            this.cachedLocationPool = newPool;
```

Comment:
`cachedLocatorPool` allows the locator to resolve requests to cached locations from the pool if the zk client has a connection loss or what have you, if the locator is not configured to operate in `strict` mode.